File: wiki/versions/1.19/data/recipes/scala/src/main/scala/org/quiltmc/wiki/recipes/Recipes.scala
package org.quiltmc.wiki.recipes

import net.minecraft.item.{ItemStack, Items}
import net.minecraft.recipe.{RecipeSerializer, RecipeType}
import net.minecraft.text.Text
import net.minecraft.util.{Formatting, Identifier}
import net.minecraft.util.random.RandomGenerator
import net.minecraft.util.registry.Registry
import org.quiltmc.loader.api.ModContainer
import org.quiltmc.qsl.base.api.entrypoint.ModInitializer
import org.quiltmc.qsl.recipe.api.RecipeManagerHelper
import org.quiltmc.qsl.recipe.api.builder.VanillaRecipeBuilders

class Recipes extends ModInitializer {

    override def onInitialize(mod: ModContainer): Unit = {
        val modId = mod.metadata.id

        // @start Registration
        Registry.register(Registry.RECIPE_TYPE, Identifier(modId, "my_recipe"), Recipes.MY_RECIPE)
        Registry.register(Registry.RECIPE_SERIALIZER, Identifier(modId, "my_recipe"), Recipes.MY_RECIPE_SERIALIZER)
        // @end Registration

        // @start Static-Recipe
        RecipeManagerHelper.registerStaticRecipe(
            VanillaRecipeBuilders.shapedRecipe("IAI")  // Shaped recipe with the IAI pattern
                    .output(ItemStack(Items.CHAINMAIL_CHESTPLATE)) // That outputs a chestplate
                    .ingredient('I', Items.IRON_INGOT) // 'I' being an ingot
                    .ingredient('A', Items.APPLE) // and 'A' an apple
                    .build(Identifier(modId, "apple_iron_ingot_fun"), "")
        )
        // @end Static-Recipe

        // @start Adding
        RecipeManagerHelper.addRecipes { handler => handler.register(Identifier(modId, "random"), id =>
                VanillaRecipeBuilders.shapelessRecipe(ItemStack(
                    // We can't know which item it's going to output in advance, therefore
                    // it's impossible to create a JSON for this recipe.
                    Registry.ITEM.getRandom(RandomGenerator.createLegacy).orElseThrow.value
                ))
                        .ingredient(Items.ACACIA_BUTTON)
                        .build(id, "")
        )}
        // @end Adding

        // @start Modification
        RecipeManagerHelper.modifyRecipes { handler =>
            // Modifies the name of the output of the stone smelting recipe
            handler.getRecipe(Identifier("minecraft", "stone"), RecipeType.SMELTING)
                    .getOutput.setCustomName(Text.literal("Modified!").formatted(Formatting.RED))
            // Replaces the enchanting table recipe with different ingredients
            handler.replace(VanillaRecipeBuilders.shapedRecipe(" B ", "D#D", "###")
                    .ingredient('B', Items.WRITABLE_BOOK)
                    .ingredient('D', Items.GOLD_INGOT)
                    .ingredient('#', Items.POLISHED_BASALT)
                    .output(ItemStack(Items.ENCHANTING_TABLE))
                    .build(Identifier("minecraft", "enchanting_table"), "")
            )
        }
        // @end Modification

        // @start Removal
        RecipeManagerHelper.removeRecipes { handler =>
            handler.remove(Identifier("minecraft", "acacia_door"))
            handler.removeIf(RecipeType.BLASTING, rec => rec.getOutput.isOf(Items.GOLD_INGOT))
        }
        // @end Removal
    }
}
object Recipes {

    // @start RecipeType-Instance
    final val MY_RECIPE: RecipeType[MyRecipe] = new RecipeType[MyRecipe] {}
    // @end RecipeType-Instance
    // @start Serializer-Instance
    final val MY_RECIPE_SERIALIZER: RecipeSerializer[MyRecipe] = MyRecipe.Serializer()
    // @end Serializer-Instance
}

File: wiki/versions/1.19/data/recipes/kotlin/src/main/kotlin/org/quiltmc/wiki/recipes/Recipes.kt
package org.quiltmc.wiki.recipes

import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.recipe.RecipeSerializer
import net.minecraft.recipe.RecipeType
import net.minecraft.text.Style
import net.minecraft.util.Formatting
import net.minecraft.util.Identifier
import net.minecraft.util.random.RandomGenerator
import net.minecraft.util.registry.Registry
import org.quiltmc.loader.api.ModContainer
import org.quiltmc.qkl.wrapper.minecraft.registry.registryScope
import org.quiltmc.qkl.wrapper.minecraft.text.buildText
import org.quiltmc.qkl.wrapper.minecraft.text.literal
import org.quiltmc.qkl.wrapper.minecraft.text.styled
import org.quiltmc.qkl.wrapper.qsl.recipe.*
import org.quiltmc.qkl.wrapper.qsl.registerEvents
import org.quiltmc.qsl.base.api.entrypoint.ModInitializer

// @start RecipeType-Instance
val MY_RECIPE: RecipeType<MyRecipe> = object: RecipeType<MyRecipe> {}
// @end RecipeType-Instance
// @start Serializer-Instance
val MY_RECIPE_SERIALIZER: RecipeSerializer<MyRecipe> = MyRecipe.Companion.Serializer()
// @end Serializer-Instance

object Recipes: ModInitializer {

    override fun onInitialize(mod: ModContainer) {
        val modId = mod.metadata().id()

        // @start Registration
        registryScope(modId) {
            MY_RECIPE withPath "my_recipe" toRegistry Registry.RECIPE_TYPE
            MY_RECIPE_SERIALIZER withPath "my_recipe" toRegistry Registry.RECIPE_SERIALIZER
        }
        // @end Registration

        // @start Static-Recipe
        registerStaticRecipe(shapedRecipe(
            Identifier(modId, "apple_iron_ingot_fun"),
            "",
            "IAI", // Shaped recipe with the IAI pattern
            ItemStack(Items.CHAINMAIL_CHESTPLATE), // That outputs a chestplate
            'I' to Items.IRON_INGOT, // 'I' being an ingot
            'A' to Items.APPLE // and 'A' an apple
        ))
        // @end Static-Recipe

        registerEvents {
            // @start Adding
            onAddRecipes {
                it.register(Identifier(modId, "random")) { id ->
                    shapelessRecipe(
                        id,
                        "",
                        // We can't know which item it's going to output in advance, therefore
                        // it's impossible to create a JSON for this recipe.
                        ItemStack(Registry.ITEM.getRandom(RandomGenerator.createLegacy()).orElseThrow().value()),
                        Items.ACACIA_BUTTON
                    )
                }
            }
            // @end Adding

            // @start Modification
            onModifyRecipes {
                // Modifies the name of the output of the stone smelting recipe
                it.getRecipe(Identifier("minecraft", "stone"), RecipeType.SMELTING)!!
                    .output.setCustomName(buildText {
                        styled(Style.EMPTY.withFormatting(Formatting.RED)) {
                            literal("Modified!")
                        }
                    })
                // Replaces the enchanting table recipe with different ingredients
                it.replace(shapedRecipe(
                    Identifier("minecraft", "enchanting_table"),
                    "",
                    """
                    | B
                    |D#D
                    |###
                    """.trimMargin(),
                    ItemStack(Items.ENCHANTING_TABLE),
                    'B' to Items.WRITABLE_BOOK,
                    'D' to Items.GOLD_INGOT,
                    '#' to Items.POLISHED_BASALT
                ))
            }
            // @end Modification

            // @start Removal
            onRemoveRecipes {
                it.remove(Identifier("minecraft", "acacia_door"))
                it.removeIf(RecipeType.BLASTING) { rec ->
                    rec.output.isOf(Items.GOLD_INGOT)
                }
            }
            // @end Removal
        }
    }
}
